crates/adapters/src/format/json/schema.rs

```diff
@@ -0,0 +1,429 @@
+        let mut typ = relation_schema(schema)?;
+        typ.logical_type = Some(LogicalType {
+            name: "Key".to_string(),
+            parameters: Default::default(),
+        });
+
+        Ok(typ)
+    }
+
+    /*pub fn relation_schema_str(schema: &Relation) -> Result<String, ControllerError> {
+        Ok(serde_json::to_string(&relation_schema(schema)?).unwrap())
+    }*/
+
+    fn relation_schema(schema: &Relation) -> Result<Type, ControllerError> {
+        let mut fields = Vec::new();
+
+        for field in schema.fields.iter() {
+            fields.push(field_schema(field)?)
+        }
+
+        Ok(Type {
+            representation_type: RepresentationType::Struct { fields },
+            logical_type: None,
+        })
+    }
+
+    fn field_schema(schema: &Field) -> Result<JsonField, ControllerError> {
+        Ok(JsonField {
+            field: schema.name.clone(),
+            typ: type_schema(&schema.columntype)?,
+            optional: schema.columntype.nullable,
+        })
+    }
+
+    fn type_schema(schema: &ColumnType) -> Result<Type, ControllerError> {
+        Ok(Type {
+            representation_type: representation_type_schema(schema)?,
+            logical_type: logical_type_schema(schema)?,
+        })
+    }
+
+    fn logical_type_schema(schema: &ColumnType) -> Result<Option<LogicalType>, ControllerError> {
+        Ok(match schema.typ.to_lowercase().as_str() {
+            "time" => Some(LogicalType {
+                name: "org.apache.kafka.connect.data.Time".to_string(),
+                parameters: BTreeMap::new(),
+            }),
+            "timestamp" => Some(LogicalType {
+                name: "org.apache.kafka.connect.data.Timestamp".to_string(),
+                parameters: BTreeMap::new(),
+            }),
+            "date" => Some(LogicalType {
+                name: "org.apache.kafka.connect.data.Date".to_string(),
+                parameters: BTreeMap::new(),
+            }),
+            _ => None,
+        })
+    }
+
+    fn representation_type_schema(
+        schema: &ColumnType,
+    ) -> Result<RepresentationType, ControllerError> {
+        match schema.typ.to_lowercase().as_str() {
+            "boolean" | "bool" => Ok(RepresentationType::Boolean),
+            "varchar" | "character varying" | "char" | "character" | "string" | "text" => {
+                Ok(RepresentationType::String)
+            }
+            "tinyint" => Ok(RepresentationType::Int8),
+            "smallint" | "int2" => Ok(RepresentationType::Int16),
+            "integer" | "int" | "signed" | "int4" => Ok(RepresentationType::Int32),
+            "bigint" | "int64" => Ok(RepresentationType::Int64),
+            // This requires Kafka connector to be configure with `"decimal.handling.mode":
+            // "string"`. Other encodings would require serde to know the scale of the
+            // decimal number, which isn't preserved in Rust.
```

Comment:
not for individual values, currently.
We can arrange for a wrapper class including type information.
In fact @abhizer implemented one, but we veto-ed it because it was wasteful.
But perhaps for these ephemeral values we can do it.

Reply:
The string encoding works just fine. We may run into data formats that don't allow that option in the future, but for now we're good.